dask/dataframe/core.py

```diff
@@ -4555,14 +4556,80 @@ def sort_values(
             **kwargs,
         )
 
+    def _check_index(self, other: Any) -> Optional[Any]:
+        """Internal method to check the column passed to set_index.
+
+        Returns None if this column is already the index.
+        """
+
+        # Check other can be translated to column name or column object
+        index = other
+        if not isinstance(index, str):
+            if isinstance(index, Sequence):
+                # Accept ["a"], but not [["a"]]
+                if len(index) == 1 and (
+                    isinstance(index[0], str) or not isinstance(index[0], Sequence)
+                ):
+                    index = index[0]
+                else:
+                    raise NotImplementedError(
+                        "Dask dataframe does not yet support multi-indexes.\n"
+                        f"You tried to index with this index: {index}\n"
+                        "Indexes must be single columns only."
+                    )
+
+            # Can also be a frame
+            elif isinstance(index, DataFrame):
+                raise NotImplementedError(
+                    "Dask dataframe does not yet support multi-indexes.\n"
+                    f"You tried to index with several columns: {list(index.columns)}\n"
+                    "Indexes must be single columns only."
+                )
+
+        # Now index should be either a series...
+        if isinstance(index, Series):
+            # If it's the current index, nothing to do
+            if index._name == self.index._name:
+                return None
+            # Otherwise, should have same length than the index
+            if (len_index := len(index)) != (len_self := len(self)):
+                raise ValueError(
+                    f"Length mismatch: series passed has {len_index} rows, but data have {len_self} rows"
+                )
```

Comment:
This check looks like it's new, but if it's a reproduction of existing behavior, please feel free to ignore. Anyhow, is it possible for `len()` to eagerly compute either `self` or `index` in this situation?

Love the application of the walrus operator, it's the first I've seen in the wild!

Reply:
Probably this is checked for later or raises an error anyway when the `map_partitions(pd.set_index)` takes place, but I think it makes more sense to add it here as a quick check. The thing is that if any `Series` is accepted as the index, it must "fit" with the `DataFrame`, but since `divisions` can be manually set, it doesn't suffice with checking `known_divisions` or `divisions`.

Yes, the walrus is great for assigning a variable that you're gonna use inside the `if` (in this case just throw-away variables after reporting the error).

Reply:
Yeah, I agree that the check should be made. I'm just concerned about the possibility of `len` running an eager compute on the collection.

Reply:
I've removed this check because it ends up failing anyway, but we can recover it if needed.

Reply:
What's the failure mode? Presumably both things _should_ be the same length...

Reply:
As of now, this is what happens in Dask:

```python
>>> ddf = dask.datasets.timeseries()
>>> ddf.npartitions
30
>>> ddf.set_index(ddf.partitions[0]['x'])    # so that lengths are different
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jorge.lopez/anaconda3/envs/AML-dev/lib/python3.9/site-packages/dask/dataframe/core.py", line 4700, in set_index
    return set_index(
  File "/home/jorge.lopez/anaconda3/envs/AML-dev/lib/python3.9/site-packages/dask/dataframe/shuffle.py", line 231, in set_index
    result = set_sorted_index(df, index, drop=drop, divisions=divisions)
  File "/home/jorge.lopez/anaconda3/envs/AML-dev/lib/python3.9/site-packages/dask/dataframe/shuffle.py", line 1103, in set_sorted_index
    raise ValueError(msg)
ValueError: When doing `df.set_index(col, sorted=True, divisions=...)`, divisions indicates known splits in the index column. In this case divisions must be the same length as the existing divisions in `df`

If the intent is to repartition into new divisions after setting the index, you probably want:

`df.set_index(col, sorted=True).repartition(divisions=divisions)`
```

The problem is that computing the `len` triggers lazy computations. Do you think the overhead of computing the `len` is worth, so that the error message is much more clear?

Reply:
This check is now done only if the user provides a `Series` in `other` which is NOT one of the columns. Since the most common scenario by far is doing `.set_index(column_name)` and in that case the `len` is not computed, I'd say this is a reasonable regarding overhead.